Add guidance for debug packages (#38546)
dotnet/docs: docs/core/distribution-packaging.md

- `System.Text.Json.dll` - installed with `dotnet-runtime-[major].[minor]` package
- `System.Text.Json.pdb` - installed with `dotnet-runtime-dbg-[major].[minor]` package

In `{dotnet_root/shared/Microsoft.AspNetCore.App/<aspnetcore version>` we'd expect the following two files:

- `Microsoft.AspNetCore.Routing.dll` - installed with `aspnetcore-runtime-[major].[minor]` packages
- `Microsoft.AspNetCore.Routing.pdb` - installed with `aspnetcore-runtime-dbg-[major].[minor]` packages

Starting with .NET 8.0, all .NET debug content (PDB files), produced by source-build, is available in a tarball named `dotnet-symbols-sdk-<version>-<rid>.tar.gz`. This archive contains PDBs in sub-directories that match directory structure of .NET SDK tarball - `dotnet-symbols-<version>-<rid>.tar.gz`.

While all debug content is available in debug tarball, not all debug content is equally important. End users are mostly interested in content of `shared/Microsoft.AspNetCore.App/<aspnetcore version>` and `shared/Microsoft.NETCore.App/<runtime version>`.

SDK content, under `sdk/<sdk version>` is useful for debugging of .NET SDK toolsets.

The following lists the recommended debug packages:

> - `dotnet-sdk-dbg-[major].[minor]` - Installs debug content for a specific SDK version

Author: Fixed with https://github.com/dotnet/docs/pull/38546/commits/376550fd3ed30555b3b302fa08c8ee0675a134b9

  - **Version:** \<sdk version>
  - **Example:** dotnet-sdk-dbg-8.0
  - **Contains:** debug content for (3),(4),(18)
  - **Dependencies:** `dotnet-sdk-[major].[minor]`

- `aspnetcore-runtime-dbg-[major].[minor]` - Installs debug content for a specific ASP.NET Core runtime
  - **Version:** \<aspnetcore runtime version>
  - **Example:** aspnetcore-runtime-dbg-8.0
  - **Contains:** debug content for (6)
  - **Dependencies:** `aspnetcore-runtime-[major].[minor]`

- `dotnet-runtime-dbg-[major].[minor]` - Installs debug content for a specific runtime
  - **Version:** \<runtime version>
  - **Example:** dotnet-runtime-dbg-8.0
  - **Contains:** debug content for (5)
  - **Dependencies:** `dotnet-runtime-[major].[minor]`

> Debug tarball also contains some debug content under `packs` which represent copies of content under `shared`. In .NET layout, `packs` directory is used for building .NET applications - there are no debugging scenarios. Thus debug content under `packs`, in debug tarball, should not be packaged.

Reviewer: ```suggestion
The debug tarball also contains some debug content under `packs` which represent copies of content under `shared`. In the .NET layout, the `packs` directory is used for building .NET applications - there are no debugging scenarios, therefore the debug content under `packs`, in the debug tarball, should not be packaged.
```

> Debug tarball also contains some debug content under `packs` which represent copies of content under `shared`. In .NET layout, `packs` directory is used for building .NET applications - there are no debugging scenarios. Thus debug content under `packs`, in debug tarball, should not be packaged.

Reviewer: Would it make sense for use to strip the packs directory from the sdk debug tarball?

Author: Yes, that is probably the best option going forward. I've created the issue: https://github.com/dotnet/source-build/issues/3769

We'll fix it in 9.0 and once backported to 8.0, we can update the guidance.

Multiple `dotnet-sdk` packages may provide the same files for the `NuGetFallbackFolder`. To avoid issues with the package manager, these files should be identical (checksum, modification date, and so on).

### Debug packages

> Debug content should be packaged in debug-named packages that follow the .NET package split described above. For instance, debug content for `dotnet-sdk-[major].[minor]` package should be included in a package named `dotnet-sdk-dbg-[major].[minor]`. Debug content should be installed to the same location as the binaries.

Reviewer 2: ```suggestion
Debug content should be packaged in debug-named packages that follow the .NET package split described previously in this article. For instance, debug content for the `dotnet-sdk-[major].[minor]` package should be included in a package named `dotnet-sdk-dbg-[major].[minor]`. You should install debug content to the same location as the binaries.
```

> Few binary examples:

Reviewer 2: ```suggestion
Here are a few binary examples:
```

> Debug content should be packaged in debug-named packages that follow the .NET package split described previously in this article. For instance, debug content for the `dotnet-sdk-[major].[minor]` package should be included in a package named `dotnet-sdk-dbg-[major].[minor]`. You should install debug content to the same location as the binaries.

Reviewer 3: Any thoughts on making this more flexible to comply with additional distro conventions? Fedora, for example, seems to prefer `debug` over `dbg`:

```
$ podman run -it fedora:39
# dnf list -q '*debug*' | wc -l
219
# dnf list -q '*dbg*' | wc -l
4
```

Reviewer 4: Part of the value of these guidelines is for different distros to use the same names.

These are all "recommendations". A maintainer can still chose to do something different.

> In `{dotnet_root}/sdk/<sdk version>` directory, we'd expect the following two files:

Reviewer 2: ```suggestion
In the `{dotnet_root}/sdk/<sdk version>` directory, the following two files are expected:
```

- `dotnet.dll` - installed with `dotnet-sdk-[major].[minor]` package
- `dotnet.pdb` - installed with `dotnet-sdk-dbg-[major].[minor]` package

> In `{dotnet_root}/shared/Microsoft.NETCore.App/<runtime version>` we'd expect the following two files:

Reviewer 2: ```suggestion
In the `{dotnet_root}/shared/Microsoft.NETCore.App/<runtime version>` directory, the following two files are expected:
```

- `System.Text.Json.dll` - installed with `dotnet-runtime-[major].[minor]` package
- `System.Text.Json.pdb` - installed with `dotnet-runtime-dbg-[major].[minor]` package

> In `{dotnet_root/shared/Microsoft.AspNetCore.App/<aspnetcore version>` we'd expect the following two files:

Reviewer 2: ```suggestion
In the `{dotnet_root/shared/Microsoft.AspNetCore.App/<aspnetcore version>` directory, the following two files are expected:
```

- `Microsoft.AspNetCore.Routing.dll` - installed with `aspnetcore-runtime-[major].[minor]` packages
- `Microsoft.AspNetCore.Routing.pdb` - installed with `aspnetcore-runtime-dbg-[major].[minor]` packages

> Starting with .NET 8.0, all .NET debug content (PDB files), produced by source-build, is available in a tarball named `dotnet-symbols-sdk-<version>-<rid>.tar.gz`. This archive contains PDBs in sub-directories that match directory structure of .NET SDK tarball - `dotnet-symbols-<version>-<rid>.tar.gz`.

Reviewer 2: ```suggestion
Starting with .NET 8.0, all .NET debug content (PDB files), produced by source-build, is available in a tarball named `dotnet-symbols-sdk-<version>-<rid>.tar.gz`. This archive contains PDBs in subdirectories that match the directory structure of the .NET SDK tarball - `dotnet-symbols-<version>-<rid>.tar.gz`.
```

> While all debug content is available in debug tarball, not all debug content is equally important. End users are mostly interested in content of `shared/Microsoft.AspNetCore.App/<aspnetcore version>` and `shared/Microsoft.NETCore.App/<runtime version>`.

Reviewer 2: ```suggestion
While all debug content is available in the debug tarball, not all debug content is equally important. End users are mostly interested in the content of the `shared/Microsoft.AspNetCore.App/<aspnetcore version>` and `shared/Microsoft.NETCore.App/<runtime version>` directories.
```

> SDK content, under `sdk/<sdk version>` is useful for debugging of .NET SDK toolsets.

Reviewer 2: ```suggestion
The SDK content under `sdk/<sdk version>` is useful for debugging .NET SDK toolsets.
```

> The following lists the recommended debug packages:

Reviewer 2: ```suggestion
The following packages are the recommended debug packages:
```

- `aspnetcore-runtime-dbg-[major].[minor]` - Installs debug content for a specific ASP.NET Core runtime
  - **Version:** \<aspnetcore runtime version>
  - **Example:** aspnetcore-runtime-dbg-8.0
  - **Contains:** debug content for (6)
  - **Dependencies:** `aspnetcore-runtime-[major].[minor]`

- `dotnet-runtime-dbg-[major].[minor]` - Installs debug content for a specific runtime
  - **Version:** \<runtime version>
  - **Example:** dotnet-runtime-dbg-8.0
  - **Contains:** debug content for (5)
  - **Dependencies:** `dotnet-runtime-[major].[minor]`

The following debug package is optional:

- `dotnet-sdk-dbg-[major].[minor]` - Installs debug content for a specific SDK version
  - **Version:** \<sdk version>
  - **Example:** dotnet-sdk-dbg-8.0
  - **Contains:** debug content for (3),(4),(18)
  - **Dependencies:** `dotnet-sdk-[major].[minor]`

> The debug tarball also contains some debug content under `packs` which represent copies of content under `shared`. In the .NET layout, the `packs` directory is used for building .NET applications - there are no debugging scenarios, therefore the debug content under `packs`, in the debug tarball, should not be packaged.

Reviewer 2: ```suggestion
The debug tarball also contains some debug content under `packs`, which represents copies of content under `shared`. In the .NET layout, the `packs` directory is used for *building* .NET applications. There are no debugging scenarios, so you shouldn't package the debug content under `packs` in the debug tarball.
```
